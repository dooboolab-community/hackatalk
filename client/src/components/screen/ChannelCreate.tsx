import {
  Animated,
  FlatList,
  Image,
  ScrollView,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { Channel, User, UserConnection, UserEdge } from '../../types/graphql';
import {
  ChannelCreateFindOrCreatePrivateChannelMutation,
  ChannelCreateFindOrCreatePrivateChannelMutationResponse,
} from '../../__generated__/ChannelCreateFindOrCreatePrivateChannelMutation.graphql';
import {
  ChannelCreateFriendsPaginationQuery,
  ChannelCreateFriendsPaginationQueryVariables,
} from '../../__generated__/ChannelCreateFriendsPaginationQuery.graphql';
import { IC_CIRCLE_X, IC_NO_IMAGE } from '../../utils/Icons';
import React, { FC, ReactElement, Suspense, useMemo, useState } from 'react';
import {
  graphql,
  useLazyLoadQuery,
  useMutation,
  usePaginationFragment,
} from 'react-relay/hooks';

import { ChannelCreateFriendsQuery } from '../../__generated__/ChannelCreateFriendsQuery.graphql';
import { ChannelCreate_friends$key } from '../../__generated__/ChannelCreate_friends.graphql';
import ErroView from '../shared/ErrorView';
import { LoadingIndicator } from 'dooboo-ui';
import { MainStackNavigationProps } from '../navigation/MainStackNavigator';
import SearchTextInput from '../shared/SearchTextInput';
import UserListItem from '../shared/UserListItem';
import { getString } from '../../../STRINGS';
import produce from 'immer';
import { showAlertForError } from '../../utils/common';
import styled from 'styled-components/native';
import useDebounce from '../../hooks/useDebounce';
import { useNavigation } from '@react-navigation/native';
import { useThemeContext } from '@dooboo-ui/theme';

const ITEM_CNT = 20;

const Container = styled.SafeAreaView`
  flex: 1;
  background-color: ${({ theme }): string => theme.backgroundDark};
  flex-direction: column;
`;

const FriendThumbView = styled.View`
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 70px;
`;

interface ChannelCreate extends User {
  checked?: boolean;
}

const findOrCreatePrivateChannel = graphql`
  mutation ChannelCreateFindOrCreatePrivateChannelMutation(
    $peerUserIds: [String!]!
  ) {
    findOrCreatePrivateChannel(peerUserIds: $peerUserIds) {
      id
      name
      channelType
    }
  }
`;

const friendsQuery = graphql`
  query ChannelCreateFriendsQuery(
    $first: Int!
    $after: String
    $searchText: String
  ) {
    ...ChannelCreate_friends
      @arguments(first: $first, after: $after, searchText: $searchText)
  }
`;

const friendsFragment = graphql`
  fragment ChannelCreate_friends on Query
  @argumentDefinitions(
    first: { type: "Int!" }
    after: { type: "String" }
    searchText: { type: "String" }
  )
  @refetchable(queryName: "ChannelCreateFriendsPaginationQuery") {
    friends(first: $first, after: $after, searchText: $searchText)
      @connection(key: "ChannelCreate_friends") {
      edges {
        cursor
        node {
          id
          email
          name
          nickname
          thumbURL
          photoURL
          birthday
          gender
          phone
          statusMessage
          verified
          lastSignedIn
          isOnline
          hasBlocked
          createdAt
          updatedAt
          deletedAt
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

type FriendsFragmentProps = {
  friend: ChannelCreate_friends$key;
  scrollY: Animated.Value;
  searchArgs: ChannelCreateFriendsPaginationQueryVariables;
  selectedUsers: User[];
  setSelectedUsers: (users: User[]) => void;
};

const FriendsFragment: FC<FriendsFragmentProps> = ({
  friend,
  selectedUsers,
  setSelectedUsers,
}) => {
  const { data } = usePaginationFragment<
    ChannelCreateFriendsPaginationQuery,
    ChannelCreate_friends$key
  >(friendsFragment, friend);

  const friendEdges = useMemo(() => {
    return (
      (data.friends as UserConnection).edges?.filter(
        (x): x is NonNullable<typeof x> => x !== null,
      ) || []
    );
  }, [data]);

  const { theme } = useThemeContext();
  // const [friends, setFriends] = useState<Friend[]>(fakeFriends);

  const navigation = useNavigation();

  const removeFriend = (friendArg: User): void => {
    const nextState = produce(selectedUsers, (draftState) => {
      const index = selectedUsers.findIndex((v) => v.id === friendArg.id);

      draftState.splice(index, 1);
    });

    setSelectedUsers(nextState);
  };

  const renderFriendThumbnail = (
    friendArg: User,
    index: number,
  ): ReactElement => {
    return (
      <FriendThumbView key={friendArg.id}>
        <View
          style={{
            marginTop: 12,
            marginRight: 16,
            marginBottom: 6,
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          <Image
            style={{
              width: 60,
              height: 60,
            }}
            source={
              friendArg.thumbURL ? { uri: friendArg.thumbURL } : IC_NO_IMAGE
            }
          />
          <Text
            numberOfLines={1}
            style={{
              fontSize: 12,
              marginTop: 4,
              color: theme.fontColor,
            }}
          >
            {friendArg.nickname || friendArg.name}
          </Text>
        </View>
        <TouchableOpacity
          testID={`remove-${index}`}
          style={{
            position: 'absolute',
            top: 0,
            right: 0,
          }}
          onPress={(): void => removeFriend(friendArg)}
        >
          <Image source={IC_CIRCLE_X} style={{ width: 32, height: 32 }} />
        </TouchableOpacity>
      </FriendThumbView>
    );
  };

  const renderItem = ({
    item,
    index,
  }: {
    item: UserEdge;
    index: number;
  }): ReactElement => {
    return (
      <UserListItem
        testID={`userlist_${index}`}
        showCheckBox
        checked={selectedUsers.includes(item?.node as User)}
        // @ts-ignore
        user={item.node}
        onPress={(): void => {
          if (selectedUsers.includes(item?.node as User)) {
            const nextState = produce(selectedUsers, (draftState) => {
              const selectedUserIndex = selectedUsers.findIndex(
                (v) => v.id === item?.node?.id,
              );

              draftState.splice(selectedUserIndex, 1);
            });

            setSelectedUsers(nextState);

            return;
          }

          const nextState = produce(selectedUsers, (draftState) => {
            draftState.push(item.node as User);
          });

          setSelectedUsers(nextState);
        }}
      />
    );
  };

  return (
    <Container>
      <FlatList
        style={{ alignSelf: 'stretch' }}
        contentContainerStyle={
          friendEdges.length === 0
            ? {
                flex: 1,
                alignSelf: 'stretch',
                flexDirection: 'column',
              }
            : null
        }
        keyExtractor={(_, index): string => index.toString()}
        data={friendEdges as UserEdge[]}
        renderItem={renderItem}
        ListHeaderComponent={(): ReactElement => {
          return (
            <ScrollView
              style={{ paddingHorizontal: 24, marginBottom: 12 }}
              horizontal
            >
              {selectedUsers.map((selectedUser, i) =>
                renderFriendThumbnail(selectedUser, i),
              )}
            </ScrollView>
          );
        }}
        ListEmptyComponent={
          <ErroView
            testID="btn-error"
            title={getString('NO_FRIEND')}
            body={getString('NO_FRIEND_BODY')}
            buttonText={getString('GO_BACK')}
            onButtonPressed={(): void => navigation.goBack()}
          />
        }
      />
    </Container>
  );
};

interface ContentProps {
  scrollY: Animated.Value;
  searchArgs: ChannelCreateFriendsPaginationQueryVariables;
  selectedUsers: User[];
  setSelectedUsers: (users: User[]) => void;
}

const ContentContainer: FC<ContentProps> = ({
  searchArgs,
  scrollY,
  selectedUsers,
  setSelectedUsers,
}) => {
  const queryResponse = useLazyLoadQuery<ChannelCreateFriendsQuery>(
    friendsQuery,
    searchArgs,
    { fetchPolicy: 'store-or-network' },
  );

  return (
    <FriendsFragment
      friend={queryResponse}
      scrollY={scrollY}
      searchArgs={searchArgs}
      selectedUsers={selectedUsers}
      setSelectedUsers={setSelectedUsers}
    />
  );
};

interface ChannelCreateProps {
  navigation: MainStackNavigationProps<'ChannelCreate'>;
}

const ChannelCreate: FC<ChannelCreateProps> = (props) => {
  const { navigation } = props;
  const [searchText, setSearchText] = useState<string>('');
  const [selectedUsers, setSelectedUsers] = useState<User[]>([]);
  const debouncedText = useDebounce(searchText, 500);
  const scrollY = new Animated.Value(0);

  const [
    commitChannel,
    isChannelInFlight,
  ] = useMutation<ChannelCreateFindOrCreatePrivateChannelMutation>(
    findOrCreatePrivateChannel,
  );

  const searchArgs: ChannelCreateFriendsPaginationQueryVariables = {
    first: ITEM_CNT,
    searchText: debouncedText,
  };

  const onChangeText = (text: string): void => {
    setSearchText(text);
    scrollY.setValue(0);

    Animated.timing(scrollY, {
      useNativeDriver: true,
      toValue: 100,
      duration: 500,
    }).start();
  };

  const pressDone = (): void => {
    const userIds = selectedUsers.map((v) => v.id);

    const mutationConfig = {
      variables: {
        peerUserIds: userIds,
      },
      onCompleted: (
        response: ChannelCreateFindOrCreatePrivateChannelMutationResponse,
      ): void => {
        const channel = response.findOrCreatePrivateChannel as Channel;

        navigation.replace('Message', {
          channel,
          users: selectedUsers,
        });
      },
      onError: (error: Error): void => {
        showAlertForError(error);
      },
    };

    commitChannel(mutationConfig);
  };

  navigation.setOptions({
    headerRight: (): ReactElement => (
      <TouchableOpacity testID="touch-done" onPress={pressDone}>
        <View
          style={{
            paddingHorizontal: 16,
            paddingVertical: 8,
          }}
        >
          <Text
            style={{
              color: 'white',
              fontSize: 14,
              fontWeight: 'bold',
            }}
          >
            {getString('DONE')}
          </Text>
        </View>
      </TouchableOpacity>
    ),
  });

  return (
    <Container>
      <SearchTextInput
        testID="text-input"
        onChangeText={onChangeText}
        containerStyle={{
          marginVertical: 12,
        }}
        value={searchText}
      />
      <Suspense fallback={<LoadingIndicator />}>
        <ContentContainer
          scrollY={scrollY}
          searchArgs={searchArgs}
          selectedUsers={selectedUsers}
          setSelectedUsers={setSelectedUsers}
        />
      </Suspense>
      {isChannelInFlight ? <LoadingIndicator /> : null}
    </Container>
  );
};

export default ChannelCreate;
