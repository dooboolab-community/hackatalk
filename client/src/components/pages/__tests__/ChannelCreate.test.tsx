import 'react-native';

import {
  createMockNavigation,
  createTestElement,
} from '../../../../test/testUtils';
import {fireEvent, render} from '@testing-library/react-native';

import ChannelCreate from '../ChannelCreate';
import React from 'react';
import ReactNavigation from '@react-navigation/core';

const mockNavigation = createMockNavigation();

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual<typeof ReactNavigation>('@react-navigation/core'),
  useNavigation: () => mockNavigation,
}));

describe('Rendering', () => {
  it('renders without crashing', () => {
    const component = createTestElement(<ChannelCreate />);
    const screen = render(component);
    const json = screen.toJSON();

    expect(json).toBeTruthy();
    expect(json).toMatchSnapshot();
  });
});

describe('Interaction', () => {
  it('should change search text', () => {
    const component = createTestElement(<ChannelCreate />);
    const screen = render(component);
    const searchInput = screen.getByTestId('text-input');

    fireEvent.changeText(searchInput, 'test search');

    expect(searchInput.props.value).toEqual('test search');
  });
});
