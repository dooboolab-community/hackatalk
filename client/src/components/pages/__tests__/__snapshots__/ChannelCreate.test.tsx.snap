// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders [UserListItem] 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
            "width": "100%",
          },
          Object {
            "marginTop": 8,
          },
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        multiline={false}
        onChangeText={[Function]}
        selectionColor="white"
        style={
          Array [
            Object {
              "backgroundColor": "#151C22",
              "borderBottomLeftRadius": 4,
              "borderBottomRightRadius": 4,
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
              "color": "white",
              "paddingBottom": 12,
              "paddingLeft": 40,
              "paddingRight": 10,
              "paddingTop": 12,
              "width": "100%",
            },
            undefined,
          ]
        }
        testID="text-input"
        underlineColorAndroid="transparent"
        value=""
      />
      <Image
        source={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "process": [Function],
          }
        }
        style={
          Array [
            Object {
              "height": 20,
              "left": 30,
              "position": "absolute",
              "width": 20,
            },
            undefined,
          ]
        }
      />
    </View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Array [
          Object {
            "backgroundColor": "#000",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          undefined,
        ]
      }
    >
      <RCTScrollView
        ListEmptyComponent={
          <ErrorView
            body="You currently do not have any friends. Please search for users to add to your friend list."
            buttonText="Go back"
            onButtonPressed={[Function]}
            testID="btn-error"
            title="No friends"
          />
        }
        ListHeaderComponent={[Function]}
        contentContainerStyle={null}
        data={
          Array [
            Object {
              "cursor": "<mock-value-for-field-\\"cursor\\">",
              "node": Object {
                "__fragmentOwner": Object {
                  "cacheConfig": Object {
                    "force": true,
                  },
                  "identifier": "203c6d15d466e67649e454eec08bda32{\\"after\\":null,\\"first\\":20,\\"includeMe\\":true,\\"searchText\\":\\"\\"}",
                  "node": Object {
                    "fragment": Object {
                      "abstractKey": null,
                      "argumentDefinitions": Array [
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "after",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "first",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "includeMe",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "searchText",
                        },
                      ],
                      "kind": "Fragment",
                      "metadata": null,
                      "name": "FriendsQuery",
                      "selections": Array [
                        Object {
                          "args": Array [
                            Object {
                              "kind": "Variable",
                              "name": "after",
                              "variableName": "after",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "first",
                              "variableName": "first",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "includeMe",
                              "variableName": "includeMe",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "searchText",
                              "variableName": "searchText",
                            },
                          ],
                          "kind": "FragmentSpread",
                          "name": "MainFriend_friends",
                        },
                        Object {
                          "args": Array [
                            Object {
                              "kind": "Variable",
                              "name": "after",
                              "variableName": "after",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "first",
                              "variableName": "first",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "includeMe",
                              "variableName": "includeMe",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "searchText",
                              "variableName": "searchText",
                            },
                          ],
                          "kind": "FragmentSpread",
                          "name": "ChannelCreate_friends",
                        },
                      ],
                      "type": "Query",
                    },
                    "hash": "83e83cdfcf6f4a0573ac1e9b2624e488",
                    "kind": "Request",
                    "operation": Object {
                      "argumentDefinitions": Array [
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "first",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "after",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "searchText",
                        },
                        Object {
                          "defaultValue": null,
                          "kind": "LocalArgument",
                          "name": "includeMe",
                        },
                      ],
                      "kind": "Operation",
                      "name": "FriendsQuery",
                      "selections": Array [
                        Object {
                          "alias": null,
                          "args": Array [
                            Object {
                              "kind": "Variable",
                              "name": "after",
                              "variableName": "after",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "first",
                              "variableName": "first",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "includeMe",
                              "variableName": "includeMe",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "searchText",
                              "variableName": "searchText",
                            },
                          ],
                          "concreteType": "UserConnection",
                          "kind": "LinkedField",
                          "name": "friends",
                          "plural": false,
                          "selections": Array [
                            Object {
                              "alias": null,
                              "args": null,
                              "concreteType": "UserEdge",
                              "kind": "LinkedField",
                              "name": "edges",
                              "plural": true,
                              "selections": Array [
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "cursor",
                                  "storageKey": null,
                                },
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "concreteType": "User",
                                  "kind": "LinkedField",
                                  "name": "node",
                                  "plural": false,
                                  "selections": Array [
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "id",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "photoURL",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "name",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "nickname",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "hasBlocked",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "statusMessage",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "isFriend",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "isOnline",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "__typename",
                                      "storageKey": null,
                                    },
                                    Object {
                                      "alias": null,
                                      "args": null,
                                      "kind": "ScalarField",
                                      "name": "thumbURL",
                                      "storageKey": null,
                                    },
                                  ],
                                  "storageKey": null,
                                },
                              ],
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "concreteType": "PageInfo",
                              "kind": "LinkedField",
                              "name": "pageInfo",
                              "plural": false,
                              "selections": Array [
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "hasNextPage",
                                  "storageKey": null,
                                },
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "endCursor",
                                  "storageKey": null,
                                },
                              ],
                              "storageKey": null,
                            },
                          ],
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": Array [
                            Object {
                              "kind": "Variable",
                              "name": "after",
                              "variableName": "after",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "first",
                              "variableName": "first",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "includeMe",
                              "variableName": "includeMe",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "searchText",
                              "variableName": "searchText",
                            },
                          ],
                          "filters": Array [],
                          "handle": "connection",
                          "key": "MainFriend_friends",
                          "kind": "LinkedHandle",
                          "name": "friends",
                        },
                        Object {
                          "alias": null,
                          "args": Array [
                            Object {
                              "kind": "Variable",
                              "name": "after",
                              "variableName": "after",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "first",
                              "variableName": "first",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "includeMe",
                              "variableName": "includeMe",
                            },
                            Object {
                              "kind": "Variable",
                              "name": "searchText",
                              "variableName": "searchText",
                            },
                          ],
                          "filters": Array [
                            "searchText",
                            "includeMe",
                          ],
                          "handle": "connection",
                          "key": "ChannelCreate_friends",
                          "kind": "LinkedHandle",
                          "name": "friends",
                        },
                      ],
                    },
                    "params": Object {
                      "cacheID": "203c6d15d466e67649e454eec08bda32",
                      "id": null,
                      "metadata": Object {},
                      "name": "FriendsQuery",
                      "operationKind": "query",
                      "text": "query FriendsQuery(
  $first: Int!
  $after: String
  $searchText: String
  $includeMe: Boolean
) {
  ...MainFriend_friends_1KCjMM
  ...ChannelCreate_friends_1KCjMM
}

fragment ChannelCreate_friends_1KCjMM on Query {
  friends(first: $first, after: $after, searchText: $searchText, includeMe: $includeMe) {
    edges {
      cursor
      node {
        id
        nickname
        name
        thumbURL
        photoURL
        ...UserListItem_user
        __typename
      }
    }
    pageInfo {
      hasNextPage
      endCursor
    }
  }
}

fragment MainFriend_friends_1KCjMM on Query {
  friends(first: $first, after: $after, searchText: $searchText, includeMe: $includeMe) {
    edges {
      cursor
      node {
        id
        ...ProfileModal_user
        ...UserListItem_user
        __typename
      }
    }
    pageInfo {
      hasNextPage
      endCursor
    }
  }
}

fragment ProfileModal_user on User {
  id
  photoURL
  name
  nickname
  hasBlocked
  statusMessage
  isFriend
}

fragment UserListItem_user on User {
  id
  photoURL
  nickname
  name
  statusMessage
  isOnline
  hasBlocked
}
",
                    },
                  },
                  "variables": Object {
                    "after": null,
                    "first": 20,
                    "includeMe": true,
                    "searchText": "",
                  },
                },
                "__fragments": Object {
                  "UserListItem_user": Object {},
                },
                "__id": "user-1",
                "__isWithinUnmatchedTypeRefinement": false,
                "__typename": "User",
                "id": "user-1",
                "name": "John Doe",
                "nickname": "jdoe1234",
                "photoURL": "<mock-value-for-field-\\"photoURL\\">",
                "thumbURL": "<mock-value-for-field-\\"thumbURL\\">",
              },
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "alignSelf": "stretch",
          }
        }
        viewabilityConfigCallbackPairs={Array []}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <RCTScrollView
              horizontal={true}
              style={
                Object {
                  "marginBottom": 8,
                  "paddingHorizontal": 24,
                }
              }
            >
              <View />
            </RCTScrollView>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "width": "100%",
                  },
                  undefined,
                ]
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
                testID="peer-button"
              >
                <View
                  isMe={false}
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#151C22",
                        "borderBottomColor": "#000",
                        "borderBottomWidth": 1,
                        "borderLeftColor": "#000",
                        "borderRightColor": "#000",
                        "borderTopColor": "#000",
                        "flexDirection": "row",
                        "height": 80,
                        "justifyContent": "flex-start",
                        "paddingBottom": 0,
                        "paddingLeft": 20,
                        "paddingRight": 20,
                        "paddingTop": 0,
                      },
                      undefined,
                    ]
                  }
                  testID="userListItem-wrapper"
                >
                  <View
                    isMe={false}
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "width": 40,
                        },
                        undefined,
                      ]
                    }
                  >
                    <Image
                      isMe={false}
                      resizeMode="cover"
                      source={
                        Object {
                          "uri": "<mock-value-for-field-\\"photoURL\\">",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "borderBottomLeftRadius": 20,
                            "borderBottomRightRadius": 20,
                            "borderTopLeftRadius": 20,
                            "borderTopRightRadius": 20,
                            "height": 40,
                            "width": 40,
                          },
                          undefined,
                        ]
                      }
                    />
                    <View
                      style={
                        Object {
                          "left": 0,
                          "position": "absolute",
                          "top": 0,
                        }
                      }
                    >
                      <Text />
                    </View>
                  </View>
                  <Text
                    numberOfLines={1}
                    style={
                      Array [
                        Object {
                          "color": "white",
                          "fontSize": 14,
                          "fontWeight": "bold",
                          "marginLeft": 12,
                          "width": 200,
                        },
                        undefined,
                      ]
                    }
                  >
                    jdoe1234
                  </Text>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "height": 24,
                          "justifyContent": "flex-start",
                        },
                        Object {
                          "position": "absolute",
                          "right": 20,
                        },
                      ]
                    }
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Object {
                          "flexDirection": "row",
                          "opacity": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "borderBottomColor": "#707070",
                              "borderBottomLeftRadius": 2,
                              "borderBottomRightRadius": 2,
                              "borderBottomWidth": 2,
                              "borderLeftColor": "#707070",
                              "borderLeftWidth": 2,
                              "borderRightColor": "#707070",
                              "borderRightWidth": 2,
                              "borderTopColor": "#707070",
                              "borderTopLeftRadius": 2,
                              "borderTopRightRadius": 2,
                              "borderTopWidth": 2,
                              "height": 20,
                              "justifyContent": "center",
                              "width": 20,
                            },
                            Array [
                              undefined,
                              Object {
                                "backgroundColor": "#000",
                                "borderColor": "#515151",
                                "borderWidth": 2,
                              },
                            ],
                          ]
                        }
                      />
                    </View>
                    <Text
                      accessible={true}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Array [
                          Object {
                            "fontSize": 14,
                            "marginLeft": 8,
                          },
                          Object {
                            "color": "white",
                          },
                        ]
                      }
                      testID="undefined-nofeed"
                    >
                      
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </RCTSafeAreaView>
  </RCTSafeAreaView>
</RNCSafeAreaProvider>
`;

exports[`Rendering renders without crashing 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 0,
            "width": "100%",
          },
          Object {
            "marginTop": 8,
          },
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        multiline={false}
        onChangeText={[Function]}
        selectionColor="white"
        style={
          Array [
            Object {
              "backgroundColor": "#151C22",
              "borderBottomLeftRadius": 4,
              "borderBottomRightRadius": 4,
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
              "color": "white",
              "paddingBottom": 12,
              "paddingLeft": 40,
              "paddingRight": 10,
              "paddingTop": 12,
              "width": "100%",
            },
            undefined,
          ]
        }
        testID="text-input"
        underlineColorAndroid="transparent"
        value=""
      />
      <Image
        source={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "process": [Function],
          }
        }
        style={
          Array [
            Object {
              "height": 20,
              "left": 30,
              "position": "absolute",
              "width": 20,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "backgroundColor": "(0, 0, 0, 0)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Image
        collapsable={false}
        source={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "process": [Function],
          }
        }
        style={
          Object {
            "borderRadius": 28,
            "opacity": 0.75,
            "transform": Array [
              Object {
                "scale": 1.7,
              },
            ],
          }
        }
      />
    </View>
  </RCTSafeAreaView>
</RNCSafeAreaProvider>
`;
